errno attribute in the aiohttp.ClientConnectorError exception is lost after 3.10.0 when multiple connection attempts fail

### Describe the bug

After aiohttp version 3.10.0, the errno attribute in the aiohttp.ClientConnectorError exception has changed. 

### To Reproduce

the case is worked before 3.10.0.
```bash
pip install pytest pytest-httpserver
```

```python
import aiohttp
import pytest
import errno
from pytest_httpserver import HTTPServer

pytestmark = pytest.mark.asyncio


async def test_session_will_raise_connection_refused_error(
    httpserver: HTTPServer
):
    httpserver.expect_request("/foobar").respond_with_data("OK")
    httpserver.stop()

    session = aiohttp.ClientSession()
    try:
        await session.get(httpserver.url_for("/foobar"))
    except aiohttp.ClientConnectorError as exc:
        # check port closed error.
        assert exc.errno == errno.ECONNREFUSED
    finally:
        await session.close()
```


### Expected behavior

testcase is passed

### Logs/tracebacks

```python-traceback
None != 61

Expected :61
Actual   :None
<Click to see difference>

../../../Library/Caches/pypoetry/virtualenvs/workspace/lib/python3.11/site-packages/aiohttp/connector.py:1036: in _wrap_create_connection
    sock = await aiohappyeyeballs.start_connection(
../../../Library/Caches/pypoetry/virtualenvs/workspace/lib/python3.11/site-packages/aiohappyeyeballs/impl.py:112: in start_connection
    raise OSError(
E   OSError: Multiple exceptions: [Errno 61] Connect call failed ('::1', 50139, 0, 0), [Errno 61] Connect call failed ('127.0.0.1', 50139)

The above exception was the direct cause of the following exception:
test4.py:17: in test_session_will_raise_connection_refused_error
    await session.get(httpserver.url_for("/foobar"))
../../../Library/Caches/pypoetry/virtualenvs/workspace/lib/python3.11/site-packages/aiohttp/client.py:648: in _request
    conn = await self._connector.connect(
../../../Library/Caches/pypoetry/virtualenvs/workspace/lib/python3.11/site-packages/aiohttp/connector.py:546: in connect
    proto = await self._create_connection(req, traces, timeout)
../../../Library/Caches/pypoetry/virtualenvs/workspace/lib/python3.11/site-packages/aiohttp/connector.py:954: in _create_connection
    _, proto = await self._create_direct_connection(req, traces, timeout)
../../../Library/Caches/pypoetry/virtualenvs/workspace/lib/python3.11/site-packages/aiohttp/connector.py:1313: in _create_direct_connection
    raise last_exc
../../../Library/Caches/pypoetry/virtualenvs/workspace/lib/python3.11/site-packages/aiohttp/connector.py:1282: in _create_direct_connection
    transp, proto = await self._wrap_create_connection(
../../../Library/Caches/pypoetry/virtualenvs/workspace/lib/python3.11/site-packages/aiohttp/connector.py:1051: in _wrap_create_connection
    raise client_error(req.connection_key, exc) from exc
E   aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host localhost:50139 ssl:default [None]

During handling of the above exception, another exception occurred:
test4.py:20: in test_session_will_raise_connection_refused_error
    assert exc.errno == errno.ECONNREFUSED
E   assert None == 61
E    +  where None = ClientConnectorError(ConnectionKey(host='localhost', port=50139, is_ssl=False, ssl=True, proxy=None, proxy_auth=None, proxy_headers_hash=None), OSError("Multiple exceptions: [Errno 61] Connect call failed ('::1', 50139, 0, 0), [Errno 61] Connect call failed ('127.0.0.1', 50139)")).errno
E    +  and   61 = errno.ECONNREFUSED
```


### Python Version

```console
$ python --version
Python 3.11.8
```


### aiohttp Version

```console
$ python -m pip show aiohttp
Name: aiohttp
Version: 3.10.0
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp
Author: 
Author-email: 
License: Apache 2
Requires: aiohappyeyeballs, aiosignal, attrs, frozenlist, multidict, yarl
Required-by: aiohttp-socks, aioresponses
```


### multidict Version

```console
$ python -m pip show multidict
```


### yarl Version

```console
$ python -m pip show yarl
Name: multidict
Version: 6.0.5
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
```


### OS

macos

### Related component

Client

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

--- a/src/aiohappyeyeballs/impl.py
+++ b/src/aiohappyeyeballs/impl.py
@@ -99,20 +99,29 @@ async def start_connection(
     if sock is None:
         all_exceptions = [exc for sub in exceptions for exc in sub]
         try:
+            first_exception = all_exceptions[0]
             if len(all_exceptions) == 1:
-                raise all_exceptions[0]
+                raise first_exception
             else:
                 # If they all have the same str(), raise one.
-                model = str(all_exceptions[0])
+                model = str(first_exception)
                 if all(str(exc) == model for exc in all_exceptions):
-                    raise all_exceptions[0]
+                    raise first_exception
                 # Raise a combined exception so the user can see all
                 # the various error messages.
-                raise OSError(
-                    "Multiple exceptions: {}".format(
-                        ", ".join(str(exc) for exc in all_exceptions)
-                    )
+                msg = "Multiple exceptions: {}".format(
+                    ", ".join(str(exc) for exc in all_exceptions)
                 )
+                # If the errno is the same for all exceptions, raise
+                # an OSError with that errno.
+                if isinstance(first_exception, OSError):
+                    first_errno = first_exception.errno
+                    if all(
+                        isinstance(exc, OSError) and exc.errno == first_errno
+                        for exc in all_exceptions
+                    ):
+                        raise OSError(first_errno, msg)
+                raise OSError(msg)
         finally:
             all_exceptions = None  # type: ignore[assignment]
             exceptions = None  # type: ignore[assignment]

--- a/src/aiohappyeyeballs/impl.py
+++ b/src/aiohappyeyeballs/impl.py
@@ -73,7 +73,7 @@ async def start_connection(
         addr_infos = _interleave_addrinfos(addr_infos, interleave)
 
     sock: Optional[socket.socket] = None
-    exceptions: List[List[Exception]] = []
+    exceptions: List[List[OSError]] = []
     if happy_eyeballs_delay is None or single_addr_info:
         # not using happy eyeballs
         for addrinfo in addr_infos:
@@ -114,13 +114,12 @@ async def start_connection(
                 )
                 # If the errno is the same for all exceptions, raise
                 # an OSError with that errno.
-                if isinstance(first_exception, OSError):
-                    first_errno = first_exception.errno
-                    if all(
-                        isinstance(exc, OSError) and exc.errno == first_errno
-                        for exc in all_exceptions
-                    ):
-                        raise OSError(first_errno, msg)
+                first_errno = first_exception.errno
+                if all(
+                    isinstance(exc, OSError) and exc.errno == first_errno
+                    for exc in all_exceptions
+                ):
+                    raise OSError(first_errno, msg)
                 raise OSError(msg)
         finally:
             all_exceptions = None  # type: ignore[assignment]
@@ -131,12 +130,12 @@ async def start_connection(
 
 async def _connect_sock(
     loop: asyncio.AbstractEventLoop,
-    exceptions: List[List[Exception]],
+    exceptions: List[List[OSError]],
     addr_info: AddrInfoType,
     local_addr_infos: Optional[Sequence[AddrInfoType]] = None,
 ) -> socket.socket:
     """Create, bind and connect one socket."""
-    my_exceptions: list[Exception] = []
+    my_exceptions: list[OSError] = []
     exceptions.append(my_exceptions)
     family, type_, proto, _, address = addr_info
     sock = None

--- a/tests/test_impl.py
+++ b/tests/test_impl.py
@@ -1191,9 +1191,10 @@ async def test_ipv64_laddr_bind_fails_all_eyeballs_interleave_first__ipv6_fails(
 
 @patch_socket
 @pytest.mark.asyncio
-async def test_all_same_exception(
+async def test_all_same_exception_and_same_errno(
     m_socket: ModuleType,
 ) -> None:
+    """Test that all exceptions are the same and have the same errno."""
     mock_socket = mock.MagicMock(
         family=socket.AF_INET,
         type=socket.SOCK_STREAM,
@@ -1256,7 +1257,7 @@ async def test_all_same_exception(
     # We should get the same exception raised if they are all the same
     with mock.patch.object(loop, "sock_connect", _sock_connect), pytest.raises(
         OSError, match="all fail"
-    ):
+    ) as exc_info:
         assert (
             await start_connection(
                 addr_info,
@@ -1266,6 +1267,98 @@ async def test_all_same_exception(
             )
             == mock_socket
         )
+
+    assert exc_info.value.errno == 5
+
+    # All calls failed
+    assert create_calls == [
+        ("dead:beef::", 80, 0, 0),
+        ("dead:aaaa::", 80, 0, 0),
+        ("107.6.106.83", 80),
+    ]
+
+
+@patch_socket
+@pytest.mark.asyncio
+async def test_all_same_exception_and_with_different_errno(
+    m_socket: ModuleType,
+) -> None:
+    """Test no errno is set if all OSError have different errno."""
+    mock_socket = mock.MagicMock(
+        family=socket.AF_INET,
+        type=socket.SOCK_STREAM,
+        proto=socket.IPPROTO_TCP,
+        fileno=mock.MagicMock(return_value=1),
+    )
+    create_calls = []
+
+    def _socket(*args, **kw):
+        for attr in kw:
+            setattr(mock_socket, attr, kw[attr])
+        return mock_socket
+
+    async def _sock_connect(
+        sock: socket.socket, address: Tuple[str, int, int, int]
+    ) -> None:
+        create_calls.append(address)
+        raise OSError(len(create_calls), "all fail")
+
+    m_socket.socket = _socket  # type: ignore
+    ipv6_addr_info = (
+        socket.AF_INET6,
+        socket.SOCK_STREAM,
+        socket.IPPROTO_TCP,
+        "",
+        ("dead:beef::", 80, 0, 0),
+    )
+    ipv6_addr_info_2 = (
+        socket.AF_INET6,
+        socket.SOCK_STREAM,
+        socket.IPPROTO_TCP,
+        "",
+        ("dead:aaaa::", 80, 0, 0),
+    )
+    ipv4_addr_info = (
+        socket.AF_INET,
+        socket.SOCK_STREAM,
+        socket.IPPROTO_TCP,
+        "",
+        ("107.6.106.83", 80),
+    )
+    addr_info = [ipv6_addr_info, ipv6_addr_info_2, ipv4_addr_info]
+    local_addr_infos = [
+        (
+            socket.AF_INET6,
+            socket.SOCK_STREAM,
+            socket.IPPROTO_TCP,
+            "",
+            ("::1", 0, 0, 0),
+        ),
+        (
+            socket.AF_INET,
+            socket.SOCK_STREAM,
+            socket.IPPROTO_TCP,
+            "",
+            ("127.0.0.1", 0),
+        ),
+    ]
+    loop = asyncio.get_running_loop()
+    # We should get the same exception raised if they are all the same
+    with mock.patch.object(loop, "sock_connect", _sock_connect), pytest.raises(
+        OSError, match="all fail"
+    ) as exc_info:
+        assert (
+            await start_connection(
+                addr_info,
+                happy_eyeballs_delay=0.3,
+                interleave=2,
+                local_addr_infos=local_addr_infos,
+            )
+            == mock_socket
+        )
+
+    # No errno is set if they are all different
+    assert exc_info.value.errno is None
 
     # All calls failed
     assert create_calls == [

